library/std/src/time/monotonic.rs

```diff
@@ -0,0 +1,93 @@
+use crate::sys::time;
+
+#[inline]
+pub(super) fn monotonize(raw: time::Instant) -> time::Instant {
+    inner::monotonize(raw)
+}
+
+#[cfg(all(target_has_atomic = "64", not(target_has_atomic = "128")))]
+pub mod inner {
+    use crate::sync::atomic::AtomicU64;
+    use crate::sync::atomic::Ordering::*;
+    use crate::sys::time;
+    use crate::time::Duration;
+
+    const ZERO: time::Instant = time::Instant::zero();
+
+    // bits 30 and 31 are never used since the seconds part never exceeds 10^9
+    const UNINITIALIZED: u64 = 0xff00_0000;
+    static MONO: AtomicU64 = AtomicU64::new(UNINITIALIZED);
+
+    #[inline]
+    pub(super) fn monotonize(raw: time::Instant) -> time::Instant {
+        let delta = raw.checked_sub_instant(&ZERO).unwrap();
+        let secs = delta.as_secs();
+        // occupies no more than 30 bits (10^9 seconds)
+        let nanos = delta.subsec_nanos() as u64;
+
+        // This wraps around every 136 years (2^32 seconds).
+        // To detect backsliding we use wrapping arithmetic and declare forward steps smaller
+        // than 2^31 seconds as expected and everything else as a backslide which will be
+        // monotonized.
+        // This could be a problem for programs that call instants at intervals greater
+        // than 68 years. Interstellar probes may want to ensure that actually_monotonic() is true.
+        let packed = (secs << 32) | nanos;
+        let old = MONO.load(Relaxed);
+
+        if packed == UNINITIALIZED || packed.wrapping_sub(old) < u64::MAX / 2 {
+            MONO.store(packed, Relaxed);
+            raw
+        } else {
+            // Backslide occurred. We reconstruct monotonized time by assuming the clock will never
+            // backslide more than 2`32 seconds which means we can reuse the upper 32bits from
+            // the seconds.
+            let secs = (secs & 0xffff_ffff << 32) | old >> 32;
```

Comment:
```suggestion
            let secs = (secs & 0xffff_ffff_0000_0000) | old >> 32;
```
I suppose that's what the code was already doing but the lack of parentheses left me very confused.

Reply:
Still, this is incorrect: consider the case where backsliding happens and produces an underflow:
old @ 0x1_0000_0000 seconds
new @ 0x0_ffff_ffff seconds

The upper bits are going to end up wrong if we use the ones from the caller since it will return a time of 0x0_0000_0000 seconds.

The solution here is that instead of using a constant `ZERO` as the base time, you should use a `Lazy<Instant>` which initializes to `Instant::now` on first use. This will allow this code to work as long as the program doesn't run for more than 63 years.

This also means that you can avoid the `UNINITIALIZED` hack and just default `MONO` to a value of 0 which means 0 difference from the first measured base time.

Reply:
Hrrm, good point. But I think it should also be possible to detect the underflow on the lower half of seconds part and subtract 1 from the upper part.

I wish this was easier to test in isolation but since the `Mutex` version is static so is this one.